{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f86215a7-cf24-4c18-a724-98aee288ba42",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import tqdm, sys, os, time, logging, warnings\n",
    "from numba.core.errors import NumbaDeprecationWarning, NumbaPendingDeprecationWarning\n",
    "warnings.simplefilter('ignore', category=NumbaDeprecationWarning)\n",
    "warnings.simplefilter('ignore', category=NumbaPendingDeprecationWarning)\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import scipy.sparse as sps\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import muon as mu\n",
    "import scvi\n",
    "\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "# sc.set_figure_params(dpi=100, fontsize=10, figsize=\"8,8\", color_map='viridis', transparent=True)\n",
    "from sklearn.metrics import adjusted_rand_score as ari"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400717ff-55b1-4eb3-800b-f1053951f1bf",
   "metadata": {},
   "source": [
    "# Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5246e2fb-8cc5-4e5a-9ca1-b4e641f7caad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 258555 × 31053\n",
       "  obs:\t&#x27;class_label&#x27;, &#x27;cluster_color&#x27;, &#x27;cluster_id&#x27;, &#x27;cluster_label&#x27;, &#x27;gene.counts&#x27;, &#x27;size&#x27;, &#x27;subclass_label&#x27;, &#x27;umi.counts&#x27;, &#x27;dataset&#x27;\n",
       "  var:\t&#x27;highly_variable&#x27;\n",
       "  uns:\t&#x27;neighbors&#x27;, &#x27;umap&#x27;\n",
       "  obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;\n",
       "  obsp:\t&#x27;connectivities&#x27;, &#x27;distances&#x27;\n",
       "  2 modalities\n",
       "    scrna:\t74132 x 31053\n",
       "      obs:\t&#x27;Amp_Date&#x27;, &#x27;Amp_Name&#x27;, &#x27;Amp_PCR_cyles&#x27;, &#x27;Cell_Capture&#x27;, &#x27;Donor&#x27;, &#x27;Gender&#x27;, &#x27;Lib_Cells&#x27;, &#x27;Lib_Date&#x27;, &#x27;Lib_Name&#x27;, &#x27;Lib_PCR_cycles&#x27;, &#x27;Lib_PassFail&#x27;, &#x27;Lib_type&#x27;, &#x27;Live_Cells&#x27;, &#x27;Live_percent&#x27;, &#x27;Mean_Reads_perCell&#x27;, &#x27;Median_Genes_perCell&#x27;, &#x27;Median_UMI_perCell&#x27;, &#x27;Region&#x27;, &#x27;Replicate_Lib&#x27;, &#x27;Saturation&#x27;, &#x27;Seq_batch&#x27;, &#x27;Total_Cells&#x27;, &#x27;aggr_num&#x27;, &#x27;class_label&#x27;, &#x27;cluster_color&#x27;, &#x27;cluster_id&#x27;, &#x27;cluster_label&#x27;, &#x27;doublet.score&#x27;, &#x27;exp_component_name&#x27;, &#x27;gene.counts&#x27;, &#x27;library_id&#x27;, &#x27;mapped_reads&#x27;, &#x27;method&#x27;, &#x27;nonconf_mapped_reads&#x27;, &#x27;size&#x27;, &#x27;subclass_label&#x27;, &#x27;total.reads&#x27;, &#x27;tube_barcode&#x27;, &#x27;umi.counts&#x27;, &#x27;unmapped_reads&#x27;, &#x27;dataset&#x27;\n",
       "      var:\t&#x27;gene_id&#x27;, &#x27;highly_variable&#x27;, &#x27;means&#x27;, &#x27;dispersions&#x27;, &#x27;dispersions_norm&#x27;\n",
       "      uns:\t&#x27;hvg&#x27;, &#x27;log1p&#x27;, &#x27;neighbors&#x27;, &#x27;pca&#x27;, &#x27;umap&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;, &#x27;counts&#x27;\n",
       "      varm:\t&#x27;PCs&#x27;\n",
       "      obsp:\t&#x27;connectivities&#x27;, &#x27;distances&#x27;\n",
       "    snrna:\t184423 x 31053\n",
       "      obs:\t&#x27;Allen.class_label&#x27;, &#x27;Allen.cluster_color&#x27;, &#x27;Allen.cluster_id&#x27;, &#x27;Allen.cluster_label&#x27;, &#x27;Allen.subclass_label&#x27;, &#x27;Broad.QC.Mito&#x27;, &#x27;Broad.QC.doublet&#x27;, &#x27;Broad.passQC&#x27;, &#x27;Comb.QC&#x27;, &#x27;MALE&#x27;, &#x27;QC&#x27;, &#x27;cl&#x27;, &#x27;class_label&#x27;, &#x27;cluster&#x27;, &#x27;cluster_color&#x27;, &#x27;cluster_id&#x27;, &#x27;cluster_label&#x27;, &#x27;comb.QC&#x27;, &#x27;dataset&#x27;, &#x27;gene.counts&#x27;, &#x27;nGene&#x27;, &#x27;nUMI&#x27;, &#x27;size&#x27;, &#x27;subclass_label&#x27;, &#x27;umi.counts&#x27;\n",
       "      var:\t&#x27;gene_id&#x27;, &#x27;highly_variable&#x27;, &#x27;means&#x27;, &#x27;dispersions&#x27;, &#x27;dispersions_norm&#x27;\n",
       "      uns:\t&#x27;hvg&#x27;, &#x27;log1p&#x27;, &#x27;neighbors&#x27;, &#x27;pca&#x27;, &#x27;umap&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;, &#x27;counts&#x27;\n",
       "      varm:\t&#x27;PCs&#x27;\n",
       "      obsp:\t&#x27;connectivities&#x27;, &#x27;distances&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 258555 × 31053\n",
       "  obs:\t'class_label', 'cluster_color', 'cluster_id', 'cluster_label', 'gene.counts', 'size', 'subclass_label', 'umi.counts', 'dataset'\n",
       "  var:\t'highly_variable'\n",
       "  uns:\t'neighbors', 'umap'\n",
       "  obsm:\t'X_pca', 'X_umap'\n",
       "  obsp:\t'connectivities', 'distances'\n",
       "  2 modalities\n",
       "    scrna:\t74132 x 31053\n",
       "      obs:\t'Amp_Date', 'Amp_Name', 'Amp_PCR_cyles', 'Cell_Capture', 'Donor', 'Gender', 'Lib_Cells', 'Lib_Date', 'Lib_Name', 'Lib_PCR_cycles', 'Lib_PassFail', 'Lib_type', 'Live_Cells', 'Live_percent', 'Mean_Reads_perCell', 'Median_Genes_perCell', 'Median_UMI_perCell', 'Region', 'Replicate_Lib', 'Saturation', 'Seq_batch', 'Total_Cells', 'aggr_num', 'class_label', 'cluster_color', 'cluster_id', 'cluster_label', 'doublet.score', 'exp_component_name', 'gene.counts', 'library_id', 'mapped_reads', 'method', 'nonconf_mapped_reads', 'size', 'subclass_label', 'total.reads', 'tube_barcode', 'umi.counts', 'unmapped_reads', 'dataset'\n",
       "      var:\t'gene_id', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "      uns:\t'hvg', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "      obsm:\t'X_pca', 'X_umap', 'counts'\n",
       "      varm:\t'PCs'\n",
       "      obsp:\t'connectivities', 'distances'\n",
       "    snrna:\t184423 x 31053\n",
       "      obs:\t'Allen.class_label', 'Allen.cluster_color', 'Allen.cluster_id', 'Allen.cluster_label', 'Allen.subclass_label', 'Broad.QC.Mito', 'Broad.QC.doublet', 'Broad.passQC', 'Comb.QC', 'MALE', 'QC', 'cl', 'class_label', 'cluster', 'cluster_color', 'cluster_id', 'cluster_label', 'comb.QC', 'dataset', 'gene.counts', 'nGene', 'nUMI', 'size', 'subclass_label', 'umi.counts'\n",
       "      var:\t'gene_id', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "      uns:\t'hvg', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "      obsm:\t'X_pca', 'X_umap', 'counts'\n",
       "      varm:\t'PCs'\n",
       "      obsp:\t'connectivities', 'distances'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata = mu.read_h5mu(\"data/processed/scRNAseq_snRNAseq_filteredQC_processedPCA.h5mu\")\n",
    "scrna = mdata['scrna']\n",
    "snrna = mdata['snrna']\n",
    "\n",
    "mdata.X = sps.vstack([mdata['scrna'].X, mdata['snrna'].X])\n",
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf2aec6f-ed84-4256-9fd0-d361c56a8714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[3.7340188 , 0.        , 0.        , 0.        , 0.        ],\n",
       "        [2.933109  , 0.        , 0.        , 0.        , 0.        ],\n",
       "        [3.1178575 , 0.414911  , 0.        , 0.        , 0.        ],\n",
       "        [3.3489263 , 0.        , 0.        , 0.38774753, 0.        ],\n",
       "        [3.7994354 , 1.0549483 , 0.        , 0.        , 0.        ]],\n",
       "       dtype=float32)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.X[:5, :5].todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87a3fcc7-802a-4d1c-af8f-813f94c5278f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[3.7340188 , 0.        , 0.        , 0.        , 0.        ],\n",
       "        [2.933109  , 0.        , 0.        , 0.        , 0.        ],\n",
       "        [3.1178575 , 0.414911  , 0.        , 0.        , 0.        ],\n",
       "        [3.3489263 , 0.        , 0.        , 0.38774753, 0.        ],\n",
       "        [3.7994354 , 1.0549483 , 0.        , 0.        , 0.        ]],\n",
       "       dtype=float32)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrna.X[:5, :5].todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f27e1b10-3dec-4497-bb2e-4ba46abfa347",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[49.,  0.,  0.,  0.,  0.],\n",
       "        [14.,  0.,  0.,  0.,  0.],\n",
       "        [42.,  1.,  0.,  0.,  0.],\n",
       "        [58.,  0.,  0.,  1.,  0.],\n",
       "        [70.,  3.,  0.,  0.,  0.]], dtype=float32)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrna.obsm['counts'][:5, :5].todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e528dfb-522c-49d8-bcce-fa8c294ffc69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 74132 × 31053\n",
       "    obs: 'Amp_Date', 'Amp_Name', 'Amp_PCR_cyles', 'Cell_Capture', 'Donor', 'Gender', 'Lib_Cells', 'Lib_Date', 'Lib_Name', 'Lib_PCR_cycles', 'Lib_PassFail', 'Lib_type', 'Live_Cells', 'Live_percent', 'Mean_Reads_perCell', 'Median_Genes_perCell', 'Median_UMI_perCell', 'Region', 'Replicate_Lib', 'Saturation', 'Seq_batch', 'Total_Cells', 'aggr_num', 'class_label', 'cluster_color', 'cluster_id', 'cluster_label', 'doublet.score', 'exp_component_name', 'gene.counts', 'library_id', 'mapped_reads', 'method', 'nonconf_mapped_reads', 'size', 'subclass_label', 'total.reads', 'tube_barcode', 'umi.counts', 'unmapped_reads', 'dataset'\n",
       "    var: 'gene_id', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'hvg', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'counts'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472f2bc0-11c5-423f-a27a-f006f57e5a02",
   "metadata": {},
   "source": [
    "# SCVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57ff64b6-62dc-4299-8e32-fcfaaba0ec0f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scrna:Amp_Date</th>\n",
       "      <th>scrna:Amp_Name</th>\n",
       "      <th>scrna:Amp_PCR_cyles</th>\n",
       "      <th>scrna:Cell_Capture</th>\n",
       "      <th>scrna:Donor</th>\n",
       "      <th>scrna:Gender</th>\n",
       "      <th>scrna:Lib_Cells</th>\n",
       "      <th>scrna:Lib_Date</th>\n",
       "      <th>scrna:Lib_Name</th>\n",
       "      <th>scrna:Lib_PCR_cycles</th>\n",
       "      <th>...</th>\n",
       "      <th>snrna:nUMI</th>\n",
       "      <th>class_label</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cluster_label</th>\n",
       "      <th>gene.counts</th>\n",
       "      <th>size</th>\n",
       "      <th>subclass_label</th>\n",
       "      <th>umi.counts</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCTTCATG-1L8TX_181211_01_G12</th>\n",
       "      <td>11/29/2018</td>\n",
       "      <td>A8TX_181129_03_A03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5898.0</td>\n",
       "      <td>426003</td>\n",
       "      <td>Male</td>\n",
       "      <td>9733.0</td>\n",
       "      <td>12/11/2018</td>\n",
       "      <td>L8TX_181211_01_G12</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>42</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>7122.000000</td>\n",
       "      <td>17334.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>55812.000000</td>\n",
       "      <td>scRNA_10x_v3 A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTGAGGTC-1L8TX_181211_01_G12</th>\n",
       "      <td>11/29/2018</td>\n",
       "      <td>A8TX_181129_03_A03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5898.0</td>\n",
       "      <td>426003</td>\n",
       "      <td>Male</td>\n",
       "      <td>9733.0</td>\n",
       "      <td>12/11/2018</td>\n",
       "      <td>L8TX_181211_01_G12</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#5DDB65</td>\n",
       "      <td>41</td>\n",
       "      <td>L5 IT Tcap_1</td>\n",
       "      <td>6125.000000</td>\n",
       "      <td>7462.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>36622.000000</td>\n",
       "      <td>scRNA_10x_v3 A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACCAGCCA-1L8TX_181211_01_G12</th>\n",
       "      <td>11/29/2018</td>\n",
       "      <td>A8TX_181129_03_A03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5898.0</td>\n",
       "      <td>426003</td>\n",
       "      <td>Male</td>\n",
       "      <td>9733.0</td>\n",
       "      <td>12/11/2018</td>\n",
       "      <td>L8TX_181211_01_G12</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#52CA74</td>\n",
       "      <td>42</td>\n",
       "      <td>L5 IT Tcap_2</td>\n",
       "      <td>8322.000000</td>\n",
       "      <td>17334.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>90475.000000</td>\n",
       "      <td>scRNA_10x_v3 A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTGAACGGT-1L8TX_181211_01_G12</th>\n",
       "      <td>11/29/2018</td>\n",
       "      <td>A8TX_181129_03_A03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5898.0</td>\n",
       "      <td>426003</td>\n",
       "      <td>Male</td>\n",
       "      <td>9733.0</td>\n",
       "      <td>12/11/2018</td>\n",
       "      <td>L8TX_181211_01_G12</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#5DDB65</td>\n",
       "      <td>41</td>\n",
       "      <td>L5 IT Tcap_1</td>\n",
       "      <td>8681.000000</td>\n",
       "      <td>7462.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>98226.000000</td>\n",
       "      <td>scRNA_10x_v3 A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTGGCATCC-1L8TX_181211_01_G12</th>\n",
       "      <td>11/29/2018</td>\n",
       "      <td>A8TX_181129_03_A03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5898.0</td>\n",
       "      <td>426003</td>\n",
       "      <td>Male</td>\n",
       "      <td>9733.0</td>\n",
       "      <td>12/11/2018</td>\n",
       "      <td>L8TX_181211_01_G12</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GABAergic</td>\n",
       "      <td>#B09FFF</td>\n",
       "      <td>20</td>\n",
       "      <td>Vip Gpc3</td>\n",
       "      <td>8081.000000</td>\n",
       "      <td>314.0</td>\n",
       "      <td>Vip</td>\n",
       "      <td>74567.000000</td>\n",
       "      <td>scRNA_10x_v3 A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>20193.0</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#3CBC92</td>\n",
       "      <td>53</td>\n",
       "      <td>L6 IT Sulf1_1</td>\n",
       "      <td>5218.066616</td>\n",
       "      <td>3978.0</td>\n",
       "      <td>L6 IT</td>\n",
       "      <td>18064.753140</td>\n",
       "      <td>snRNA_10x_v3 B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2858.0</td>\n",
       "      <td>Non-Neuronal</td>\n",
       "      <td>#474662</td>\n",
       "      <td>90</td>\n",
       "      <td>Oligo Opalin_4</td>\n",
       "      <td>1740.348364</td>\n",
       "      <td>16566.0</td>\n",
       "      <td>Oligo</td>\n",
       "      <td>3419.937704</td>\n",
       "      <td>snRNA_10x_v3 B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>35854.0</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#00CF1E</td>\n",
       "      <td>51</td>\n",
       "      <td>L5 IT S100b</td>\n",
       "      <td>5464.036043</td>\n",
       "      <td>8684.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>20579.022920</td>\n",
       "      <td>snRNA_10x_v3 B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>23493.0</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#3CBC92</td>\n",
       "      <td>52</td>\n",
       "      <td>L5 IT Pld5</td>\n",
       "      <td>5680.692074</td>\n",
       "      <td>3621.0</td>\n",
       "      <td>L5 IT</td>\n",
       "      <td>22718.304610</td>\n",
       "      <td>snRNA_10x_v3 B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>22385.0</td>\n",
       "      <td>Glutamatergic</td>\n",
       "      <td>#6F836B</td>\n",
       "      <td>48</td>\n",
       "      <td>L2/3 IT_1</td>\n",
       "      <td>5300.903467</td>\n",
       "      <td>23971.0</td>\n",
       "      <td>L2/3 IT</td>\n",
       "      <td>19543.896210</td>\n",
       "      <td>snRNA_10x_v3 B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>258555 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             scrna:Amp_Date   \n",
       "sample_name                                                   \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12             11/29/2018  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12             11/29/2018   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12             11/29/2018   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12             11/29/2018   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12             11/29/2018   \n",
       "...                                                     ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG            NaN   \n",
       "\n",
       "                                                  scrna:Amp_Name   \n",
       "sample_name                                                        \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12          A8TX_181129_03_A03  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12          A8TX_181129_03_A03   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12          A8TX_181129_03_A03   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12          A8TX_181129_03_A03   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12          A8TX_181129_03_A03   \n",
       "...                                                          ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG                 NaN   \n",
       "\n",
       "                                              scrna:Amp_PCR_cyles   \n",
       "sample_name                                                         \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                         12.0  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                         12.0   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                         12.0   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                         12.0   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                         12.0   \n",
       "...                                                           ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA                  NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG                  NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG                  NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC                  NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG                  NaN   \n",
       "\n",
       "                                              scrna:Cell_Capture scrna:Donor   \n",
       "sample_name                                                                    \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                      5898.0      426003  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                      5898.0      426003   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                      5898.0      426003   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                      5898.0      426003   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                      5898.0      426003   \n",
       "...                                                          ...         ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA                 NaN         NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG                 NaN         NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG                 NaN         NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC                 NaN         NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG                 NaN         NaN   \n",
       "\n",
       "                                             scrna:Gender  scrna:Lib_Cells   \n",
       "sample_name                                                                  \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                 Male           9733.0  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                 Male           9733.0   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                 Male           9733.0   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                 Male           9733.0   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                 Male           9733.0   \n",
       "...                                                   ...              ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA          NaN              NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG          NaN              NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG          NaN              NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC          NaN              NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG          NaN              NaN   \n",
       "\n",
       "                                             scrna:Lib_Date   \n",
       "sample_name                                                   \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12             12/11/2018  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12             12/11/2018   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12             12/11/2018   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12             12/11/2018   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12             12/11/2018   \n",
       "...                                                     ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC            NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG            NaN   \n",
       "\n",
       "                                                  scrna:Lib_Name   \n",
       "sample_name                                                        \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12          L8TX_181211_01_G12  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12          L8TX_181211_01_G12   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12          L8TX_181211_01_G12   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12          L8TX_181211_01_G12   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12          L8TX_181211_01_G12   \n",
       "...                                                          ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC                 NaN   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG                 NaN   \n",
       "\n",
       "                                              scrna:Lib_PCR_cycles  ...   \n",
       "sample_name                                                         ...   \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                           9.0  ...  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                           9.0  ...   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                           9.0  ...   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                           9.0  ...   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                           9.0  ...   \n",
       "...                                                            ...  ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA                   NaN  ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG                   NaN  ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG                   NaN  ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC                   NaN  ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG                   NaN  ...   \n",
       "\n",
       "                                             snrna:nUMI    class_label   \n",
       "sample_name                                                              \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                NaN  Glutamatergic  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                NaN  Glutamatergic   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                NaN  Glutamatergic   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                NaN  Glutamatergic   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                NaN      GABAergic   \n",
       "...                                                 ...            ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA    20193.0  Glutamatergic   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG     2858.0   Non-Neuronal   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG    35854.0  Glutamatergic   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC    23493.0  Glutamatergic   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG    22385.0  Glutamatergic   \n",
       "\n",
       "                                              cluster_color cluster_id   \n",
       "sample_name                                                              \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12                #52CA74         42  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12                #5DDB65         41   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12                #52CA74         42   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12                #5DDB65         41   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                #B09FFF         20   \n",
       "...                                                     ...        ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA        #3CBC92         53   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG        #474662         90   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG        #00CF1E         51   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC        #3CBC92         52   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG        #6F836B         48   \n",
       "\n",
       "                                               cluster_label  gene.counts   \n",
       "sample_name                                                                 \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12            L5 IT Tcap_2  7122.000000  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12            L5 IT Tcap_1  6125.000000   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12            L5 IT Tcap_2  8322.000000   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12            L5 IT Tcap_1  8681.000000   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12                Vip Gpc3  8081.000000   \n",
       "...                                                      ...          ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA   L6 IT Sulf1_1  5218.066616   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG  Oligo Opalin_4  1740.348364   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG     L5 IT S100b  5464.036043   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC      L5 IT Pld5  5680.692074   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG       L2/3 IT_1  5300.903467   \n",
       "\n",
       "                                                 size subclass_label   \n",
       "sample_name                                                            \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12          17334.0          L5 IT  \\\n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12           7462.0          L5 IT   \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12          17334.0          L5 IT   \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12           7462.0          L5 IT   \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12            314.0            Vip   \n",
       "...                                               ...            ...   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA   3978.0          L6 IT   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG  16566.0          Oligo   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG   8684.0          L5 IT   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC   3621.0          L5 IT   \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG  23971.0        L2/3 IT   \n",
       "\n",
       "                                                umi.counts         dataset  \n",
       "sample_name                                                                 \n",
       "AAACCCAAGCTTCATG-1L8TX_181211_01_G12          55812.000000  scRNA_10x_v3 A  \n",
       "AAACCCAAGTGAGGTC-1L8TX_181211_01_G12          36622.000000  scRNA_10x_v3 A  \n",
       "AAACCCACACCAGCCA-1L8TX_181211_01_G12          90475.000000  scRNA_10x_v3 A  \n",
       "AAACCCAGTGAACGGT-1L8TX_181211_01_G12          98226.000000  scRNA_10x_v3 A  \n",
       "AAACCCAGTGGCATCC-1L8TX_181211_01_G12          74567.000000  scRNA_10x_v3 A  \n",
       "...                                                    ...             ...  \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTCATGAGTAA  18064.753140  snRNA_10x_v3 B  \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGGTTTCGCAAGAG   3419.937704  snRNA_10x_v3 B  \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGAGACTCTTG  20579.022920  snRNA_10x_v3 B  \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCACCTTGC  22718.304610  snRNA_10x_v3 B  \n",
       "pBICCNsMMrMOpRMiM006d190320_TTTGTTGTCATGCTAG  19543.896210  snRNA_10x_v3 B  \n",
       "\n",
       "[258555 rows x 57 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1a972bf-642a-479e-9c39-6453d6b28191",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Detected unpaired observations in modality scrna. Please make sure that data is fully paired in all MuData inputs. Either pad the unpaired modalities or take the intersection with muon.pp.intersect_obs().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mscvi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSCVI\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msetup_anndata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlayer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbatch_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msample_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlabels_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msubclass_label\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.9/site-packages/scvi/model/_scvi.py:211\u001b[0m, in \u001b[0;36mSCVI.setup_anndata\u001b[0;34m(cls, adata, layer, batch_key, labels_key, size_factor_key, categorical_covariate_keys, continuous_covariate_keys, **kwargs)\u001b[0m\n\u001b[1;32m    207\u001b[0m     anndata_fields \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_get_fields_for_adata_minification(adata_minify_type)\n\u001b[1;32m    208\u001b[0m adata_manager \u001b[38;5;241m=\u001b[39m AnnDataManager(\n\u001b[1;32m    209\u001b[0m     fields\u001b[38;5;241m=\u001b[39manndata_fields, setup_method_args\u001b[38;5;241m=\u001b[39msetup_method_args\n\u001b[1;32m    210\u001b[0m )\n\u001b[0;32m--> 211\u001b[0m \u001b[43madata_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mregister_fields\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    212\u001b[0m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mregister_manager(adata_manager)\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.9/site-packages/scvi/data/_manager.py:171\u001b[0m, in \u001b[0;36mAnnDataManager.register_fields\u001b[0;34m(self, adata, source_registry, **transfer_kwargs)\u001b[0m\n\u001b[1;32m    166\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m source_registry \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m transfer_kwargs:\n\u001b[1;32m    167\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    168\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregister_fields() got unexpected keyword arguments \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtransfer_kwargs\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m passed without a source_registry.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    169\u001b[0m     )\n\u001b[0;32m--> 171\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_anndata_object\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m field \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfields:\n\u001b[1;32m    174\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_add_field(\n\u001b[1;32m    175\u001b[0m         field\u001b[38;5;241m=\u001b[39mfield,\n\u001b[1;32m    176\u001b[0m         adata\u001b[38;5;241m=\u001b[39madata,\n\u001b[1;32m    177\u001b[0m         source_registry\u001b[38;5;241m=\u001b[39msource_registry,\n\u001b[1;32m    178\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mtransfer_kwargs,\n\u001b[1;32m    179\u001b[0m     )\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.9/site-packages/scvi/data/_manager.py:111\u001b[0m, in \u001b[0;36mAnnDataManager._validate_anndata_object\u001b[0;34m(self, adata)\u001b[0m\n\u001b[1;32m    105\u001b[0m     _check_if_view(adata, copy_if_view\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    107\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    108\u001b[0m     \u001b[38;5;28misinstance\u001b[39m(adata, MuData)\n\u001b[1;32m    109\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalidation_checks\u001b[38;5;241m.\u001b[39mcheck_fully_paired_mudata\n\u001b[1;32m    110\u001b[0m ):\n\u001b[0;32m--> 111\u001b[0m     \u001b[43m_check_mudata_fully_paired\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/lib/python3.9/site-packages/scvi/data/_utils.py:247\u001b[0m, in \u001b[0;36m_check_mudata_fully_paired\u001b[0;34m(mdata)\u001b[0m\n\u001b[1;32m    245\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m mod_key \u001b[38;5;129;01min\u001b[39;00m mdata\u001b[38;5;241m.\u001b[39mmod:\n\u001b[1;32m    246\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m mdata\u001b[38;5;241m.\u001b[39mobsm[mod_key]\u001b[38;5;241m.\u001b[39mall():\n\u001b[0;32m--> 247\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    248\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDetected unpaired observations in modality \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmod_key\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    249\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPlease make sure that data is fully paired in all MuData inputs. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    250\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEither pad the unpaired modalities or take the intersection with muon.pp.intersect_obs().\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    251\u001b[0m         )\n",
      "\u001b[0;31mValueError\u001b[0m: Detected unpaired observations in modality scrna. Please make sure that data is fully paired in all MuData inputs. Either pad the unpaired modalities or take the intersection with muon.pp.intersect_obs()."
     ]
    }
   ],
   "source": [
    "scvi.model.SCVI.setup_anndata(mdata, \n",
    "                              layer='counts', \n",
    "                              batch_key=\"sample_name\", \n",
    "                              labels_key='subclass_label')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7883b1e5-f88e-45ca-954e-63bbaedb1204",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
